src/daemon/distribution.c

```diff
@@ -0,0 +1,205 @@
+  build_tree(d, buckets, left_child, left, mid);
+  build_tree(d, buckets, right_child, mid + 1, right);
+  d->tree[node_index] = merge_buckets(d->tree[left_child], d->tree[right_child]);
+}
+
+distribution_t* build_distribution_from_bucket_array(size_t num_buckets, bucket_t *bucket_array) {
+  distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+  if (new_distribution == NULL) {
+    return NULL;
+  }
+  new_distribution->tree = calloc(2 * num_buckets - 1, sizeof(bucket_t));
+  if (new_distribution->tree == NULL) {
+    free(new_distribution);
+    return NULL;
+  }
+
+  new_distribution->num_buckets = num_buckets;
+  build_tree(new_distribution, bucket_array, 0, 0, num_buckets - 1);
+  return new_distribution;
+}
+
+distribution_t* distribution_new_linear(size_t num_buckets, double size) {
+  if (num_buckets == 0 || size <= 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  bucket_t *bucket_array = calloc(num_buckets, sizeof(bucket_t));
+  if (bucket_array == NULL)
+    return NULL;
+
+  for (size_t i = 0; i < num_buckets; i++) {
+    bucket_array[i].minimum = i * size;
+    if (i == num_buckets - 1)
+      bucket_array[i].maximum = INFINITY;
+    else
+      bucket_array[i].maximum = (i + 1) * size;
+  }
+
+  distribution_t *new_distribution = build_distribution_from_bucket_array(num_buckets, bucket_array);
+  free(bucket_array);
+  return new_distribution;
+}
+
+distribution_t* distribution_new_exponential(size_t num_buckets, double initial_size, double factor) {
+  if (num_buckets == 0 || initial_size <= 0 || factor <= 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  bucket_t *bucket_array = calloc(num_buckets, sizeof(bucket_t));
+  if (bucket_array == NULL)
+    return NULL;
+
+  bucket_array[0] = (bucket_t) {
+    .maximum = initial_size,
+  };
+  for (size_t i = 1; i < num_buckets; i++) {
+    bucket_array[i].minimum = bucket_array[i - 1].maximum;
+    if (i == num_buckets - 1)
+      bucket_array[i].maximum = INFINITY;
+    else
+      bucket_array[i].maximum = bucket_array[i].minimum * factor;
+  }
+
+  distribution_t *new_distribution = build_distribution_from_bucket_array(num_buckets, bucket_array);
+  free(bucket_array);
+  return new_distribution;
+}
+
+distribution_t* distribution_new_custom(size_t num_buckets, double *custom_buckets_sizes) {
+  if (num_buckets == 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+  for (size_t i = 0; i < num_buckets; i++)
+    if (custom_buckets_sizes[i] <= 0) {
+      errno = EINVAL;
+      return NULL;
+    }
+
+  bucket_t *bucket_array = calloc(num_buckets + 1, sizeof(bucket_t));
+  if (bucket_array == NULL)
+    return NULL;
+
+  for (size_t i = 0; i < num_buckets + 1; i++) {
+    if (i != 0)
+      bucket_array[i].minimum = bucket_array[i - 1].maximum;
+    if (i == num_buckets)
+      bucket_array[i].maximum = INFINITY;
+    else
+      bucket_array[i].maximum = bucket_array[i].minimum + custom_buckets_sizes[i];
```

Comment:
I'd argue that providing a list of bucket boundaries is much easier to read than providing bucket sizes. Consider:

```c
double boundaries[] = {1, 2, 5, 10, 20, 50, 100};
```

vs.

```c
double sizes[] = {1, 1, 3, 5, 10, 30, 50};
```